### DISCUSSION TITLE
[bug] USB power trigger does not persist

### DISCUSSION THREAD

**Discussion by salarmehr** [gopro/labs#271] [General]
<img width="750" alt="image" src="https://user-images.githubusercontent.com/2582036/226152534-3a7a7268-2253-4b22-8752-0bf27ca9f2eb.png">

Using the above code, I activated the USB power trigger to use the camera like a dashcam. It works until I turn on the camera with the power button or if I switch it off with the power button or capture anything manually. The trigger does not work anymore until I scan the code again. 

This means that I cannot use the camera to capture something and put it back on the car to be used like a dashcam and I have to scan the QR code again and again. 

I hope in the future you can simply add a new mode next to slow motion mode, to use the camera like a dash cam with a cute car icon!

**Comment by dnewman-gpsw** [comment]
That example works as designed, 3 years ago. So not a bug.  If you want a more modern way you need to use the WAKE and BOOT commands. see https://gopro.github.io/labs/control/extensions/ This way you can either manually or automatically boot the camera, and have it perform any action you want.  In the latest HERO11 Labs firmware, you can do even more to customize a dashcam mode, using Timewarp when the car is not in motion, or really any combination of camera features you can imagine.

**Reply by TripleFJeff** [reply]
Is "WAKE" the same as turning the camera on, and should it work without a boot command? New at using all of this; as an experiment I did a simple !Mwake=2. I know that the QR "worked" in that I see the wake=2 setting displayed every time I turn the camera on. But when I plug it into USB power it doesn't turn on. I do get the red status light, but from what I can tell that just means it is charging? This is on a Hero 8 Black with the latest labs firmware.

**Reply by dnewman-gpsw** [reply]
"wake" is not the same as "WAKE", all commands are case sensitive.

**Reply by TripleFJeff** [reply]
That explains it, thank you. I noticed a few odd behaviors once I properly added the WAKE=2:

When I turn the camera on manually, I get the "Overlays Active" and "Features activated" screen. If it wakes on power, I only get the "Overlays Active" screen, although it still functions correctly. Is this normal?

If the camera is on USB power but off, if I remove power it turns on. Is this supposed to happen?

Related question- is there a way to perform an action when USB power is removed? In particular I would like to wake the camera on USB power and keep it on by setting the auto off to never. But once power is removed, I want to change the auto off to some other value so it doesn't stay on until the battery dies.

**Reply by dnewman-gpsw** [reply]
I don't know want else you have active in the camera, reporting "Overlay Active" is likely correct for your setup.

The power is booting on any power change with WAKE=2.  

Yes, you can script all that behavior fairly easily, but only with HERO11 and 11-Mini.   With HERO8, you have to stick with the less reliable https://gopro.github.io/labs/control/usb/ Then experiment with that script.

**Reply by TripleFJeff** [reply]
The part I thought was odd was that with no change in setup, the startup message was different depending on how it was turned on-- when I use the power button I get both, but when it wakes on power I don't get the "Features activated" screen. But function seems to be the same.

If I understand correctly, using the WAKE=2 will cause the camera to turn on with USB power but any boot script I have runs regardless of how it is started-- USB or with power button. And since the 8 can't use the conditionals based on status I have no way to implement anything that depends on USB power status. The USB power trigger function allows me to run a command or script only if it is turned on via USB and the only thing it can do at USB power off is to "end" the command. If my USB Power Trigger command is oC0 (never auto off), will that command be "undone" at power off, or are the settings more of a one-shot change that can't be "ended" like a command?

**Reply by dnewman-gpsw** [reply]
This is where you need to experiment.  There might be a combination that works for your needs. As for overlay active, vs feature activated, that depends on what you are going, which I have no idea.  Million of combinations.  Labs has so many hacks, that can be stacked in interesting ways.

**Reply by TripleFJeff** [reply]
I'm using it as a "dashcam" in a race car with the GPS speed start/stop. I want it to wake on USB power and never turn off as long as it is powered. I was hoping to find a way to either turn it off or switch the auto-off setting when it loses USB power so I don't kill the battery, but maybe it doesn't matter? Maybe I don't actually need to run a battery at all? As long as I don't come to a stop and turn off the master switch before the hold time is up and it closes the file I shouldn't need a battery.

**Reply by dnewman-gpsw** [reply]
Simple solution:  Not use a battery, upon wake auto-start recording with a boot command.  When power goes out, the capture is abruptly stopped, not worry, it will be auto-repaired on the next boot.  The whole thing repeats on power.

**Reply by TripleFJeff** [reply]
For anyone following along, a few things that I have found:
-When the start on speed function is being used (!SK) it will keep the camera on regardless of any auto power off setting. If you manually cancel it using the shutter button, it will auto power off based on the setting.
-The USB power trigger (!u) can be used to start the !SK function, but removing power will not stop it like it will with the !S function
-It seems that with an 8 (that doesn't have the ability to do conditionals based on status), once a start on speed has been initiated there is no way to automatically stop and/or power it off based on usb power

Since I only want to record while moving, I think the best option is no battery, WAKE=2, and use a boot script to put it into start-on-speed, where it will stay until it loses power. I could keep a battery in there, but it is always going to kill it eventually.

**Comment by salarmehr** [comment]
@dnewman-gpsw thanks, could you advise what is the equivalent command (Metadata Value) for "Trigger on USB" but just to keep it permanent? (so you can use is a dash cam and occasionally capture some videos manually? (when powered on by pressing a button it can be used as a camera and when powered up because of USB it can continue working like a dash cam. Thanks.

**Comment by salarmehr** [comment]
@dnewman-gpsw additionally, Consider I switch off the car and the camera will turn off after one minute because of the setting, meanwhile, it could connect to the home WiFi. Is it possible to set the camera to first upload the videos first and turn off after the upload is done?
Thanks.

**Comment by dnewman-gpsw** [comment]
Once you can clarify want you want some of this might be possible, but you need to create it for your custom needs.

The basics are in this format !MWAKE=2!MBOOT="!Lboot"!SAVEboot=your_command_here

For a dashcam this is an example.  !MWAKE=2!MBOOT="!Lboot"!SAVEboot=">u0!S+!1N=Bt:B>u0=AB=CB=C-A>C10!E+!1O>r0!R9"

For Coders:  
```
   >u0!S+!1N     // if(usb_power > 0) StartCapture(), Nop(1 second);
loop:
   =Bt:B         // B =  TimeSinceBootInSeconds();  
   >u0=AB        // if(usb_power > 0)  A=B;
   =CB           // C=B;
   =C-A          // C=C-A;
   >C10!E+!1O    // if(C > 10) EndCapture, ShutdownSeconds(1);
   >r0!R9        // if(recording) goto loop;
```

**Comment by salarmehr** [comment]
@dnewman-gpsw do you have a link to the coding documentation? Is this a general syntax or in-house developed by GoPro?

**Reply by dnewman-gpsw** [reply]
This is the GitHub for the documentation.  All here , pages of it, https://github.com/gopro/labs

It is GoPro custom, for compactness, as it is all scripted within limited bandwidth QR codes.

**Comment by mitts-hub** [comment]
@salarmehr did you ever find a solution for this? I am trying to set up my gopros in myh racecars as well but running into the same issue. Any help would be greatly appreciated.

**Reply by dnewman-gpsw** [reply]
This is a lot of info here., including solutions.  What is not working for you?

**Reply by mitts-hub** [reply]
@dnewman-gpsw 

I am trying this command  !MWAKE=2!MBOOT="!Lboot"!SAVEboot=!uN!!u30E!R"
I am generating the QR code at the bottom of the page here:
https://gopro.github.io/labs/control/actions/

Any help would be greatly appreciated.

**Reply by dnewman-gpsw** [reply]
There are typos in that command !MWAKE=2!MBOOT="!Lboot"!SAVEboot=!uN!u30E!R

**Reply by mitts-hub** [reply]
@dnewman-gpsw

Thank you I took your modified code and changed it to !MWAKE=2!MBOOT="!Lboot"!SAVEboot=!uN!S!u30E!R

This should trigger the camera to turn on and start recording in the default video mode when USB power is enabled. Instead the camera just turns on then says "shutdown until usb power" even though power is being applied. The good news is this command persists as I can turn the camera on battery only (I get the same shutdown until usb power message) and then get the same message when I apply usb power.

**Reply by dnewman-gpsw** [reply]
Of course, when combining new and old techniques just do !MWAKE=2!MBOOT="!Lboot"!SAVEboot=!S!u30E!1O

**Reply by mitts-hub** [reply]
@dnewman-gpsw odd, now the camera will only charge on USB power. If I turn on the camera under battery power it will start recording but think that that the USB power is disconnected and record for the 30 seconds.

**Reply by dnewman-gpsw** [reply]
We are trying for the old solution for older cameras. You didn't mention what camera you are running, as all can runs well on HERO11, with the newer command language.

**Reply by mitts-hub** [reply]
@dnewman-gpsw Sorry, running a GoPro Hero 9 Black

**Reply by dnewman-gpsw** [reply]
That will be the issue, as WAKE=2 is not fixed on HERO9 listed compatibility is: HERO8/10/11 only. I just tried on the HERO9, and it will not wake with !MWAKE=2. So you can't do boot command with the old USB trigger.   However the sample code on https://gopro.github.io/labs/control/usb/ does work on HERO9, with its limitations.

**Reply by TripleFJeff** [reply]
Having just implemented this for a race car with a HERO 8, some info that might be useful.

I'm using the WAKE=2, but my process for setting up the camera has been to set the WAKE=2 in a separate action than the !MBOOT. I don't know if that has any relevance, but could you simply substitute the old USB trigger for the WAKE=2, but with no action (or some nominal action if required like setting mode to video)? Or does the !MBOOT not run when the old USB trigger wakes the camera?

Regardless, I think an easier way to go would be to use a trigger based on either GPS speed or IMU to start/stop recording combined with setting the Auto Off to Never. That way you are only recording when it is moving, and when the car powers off the camera just stays on until the battery dies. The idle time on battery is actually pretty long, so as long as you have some time where it is powered but not recording to keep the battery topped up, it will probably stay on all day and not die until overnight. What kind of racing are you using it for?

For reference, this is my process for setting up a camera. I have these in a spreadsheet and paste them into the QR generator and scan in order.
!RESET (Reset)
g1!MBRNO=10!MBURN="(0,40)[BCHH:MM:SS AA mm-dd-yyyy ]" (Set Time/Date Burn in)
mVr1080p60fSoSM25,50t0e1hS0aSg1q0dVoV0B4D4R0CS1 (Basic Settings)
!MWAKE=2 (Wake on power)
!MBOOT="!Lspd"!SAVEspd=!SK32H30!R (Boot command(Speed Capture 32kph 30s))
